pkg/controller/networkpolicy/mutate.go

```diff
@@ -0,0 +1,201 @@
+		}
+		if oldRaw != nil {
+			if err := json.Unmarshal(oldRaw, &oldACNP); err != nil {
+				klog.Errorf("Error de-serializing old Antrea ClusterNetworkPolicy")
+				return GetAdmissionResponseForErr(err)
+			}
+		}
+		msg, allowed, patch = m.mutateAntreaPolicyRuleName(op, curACNP.Spec.Ingress, curACNP.Spec.Egress)
+	case "NetworkPolicy":
+		klog.V(2).Info("Mutating Antrea NetworkPolicy CRD")
+		var curANP, oldANP secv1alpha1.NetworkPolicy
+		if curRaw != nil {
+			if err := json.Unmarshal(curRaw, &curANP); err != nil {
+				klog.Errorf("Error de-serializing current Antrea NetworkPolicy")
+				return GetAdmissionResponseForErr(err)
+			}
+		}
+		if oldRaw != nil {
+			if err := json.Unmarshal(oldRaw, &oldANP); err != nil {
+				klog.Errorf("Error de-serializing old Antrea NetworkPolicy")
+				return GetAdmissionResponseForErr(err)
+			}
+		}
+		msg, allowed, patch = m.mutateAntreaPolicyRuleName(op, curANP.Spec.Ingress, curANP.Spec.Egress)
+	}
+
+	if msg != "" {
+		result = &metav1.Status{
+			Message: msg,
+		}
+	}
+	return &admv1.AdmissionResponse{
+		Allowed:   allowed,
+		Result:    result,
+		PatchType: &patchType,
+		Patch:     patch,
+	}
+}
+
+// mutateAntreaPolicyRuleName mutates names of rules of an Antrea NetworkPolicy CRD.
+// If users didn't specify the name of an ingress or egress rule,
+// mutateAntreaPolicyRuleName will auto-generate a name for this rule.
+func (m *NetworkPolicyMutator) mutateAntreaPolicyRuleName(op admv1.Operation, ingress, egress []secv1alpha1.Rule) (string, bool, []byte) {
+	allowed := true
+	reason := ""
+	var patch []byte
+	switch op {
+	case admv1.Create, admv1.Update:
+		ruleNames := sets.NewString()
+		for _, rule := range append(ingress, egress...) {
+			ruleNames.Insert(rule.Name)
+		}
+
+		ingressPath, ingressValue := generateName(ruleNames, "ingress", ingress)
+		egressPath, egressValue := generateName(ruleNames, "egress", egress)
+
+		genPatch, err := createReplacePatch(append(ingressPath, egressPath...), append(ingressValue, egressValue...))
+		if err != nil {
+			allowed = false
+			reason = "unable to generate mutating patch"
+			break
+		}
+		patch = genPatch
+
+	case admv1.Delete:
+		// Delete of Antrea Policies have no mutation
+		allowed = true
+	}
+	return reason, allowed, patch
+}
+
+// generateName generate unique rule name and return a list of jsonPath and a list generated name
+func generateName(ruleNames sets.String, prefix string, rules []secv1alpha1.Rule) ([]string, []string) {
+	var paths []string
+	var values []string
+	for idx, rule := range rules {
+		if rule.Name == "" {
+			genName := fmt.Sprintf("%s-%s-%s", prefix, strings.ToLower(string(*rule.Action)), hashRule(rule))
+			// If an existing name is generated, it means there are duplicate rules or
+			// one of rule-names contains hash value of other rules, which is illegal.
+			// MutatingWebhook will skip auto-gen of this rule as a signal to pass illegal input info to ValidatingWebhook
+			if ruleNames.Has(genName) {
+				break
```

Comment:
I kind of feel like we don't need to do this here, or maintain a set of existing rule names. We can just auto-generate the rule name, and the `ValidatingWebhook` will deal with it?

Reply:
IMO, ValidatingWebhook should know the non-uniqueness is caused by user-provided names or auto-gen names to give users a precise response. So I use this "skip" action as a signal for ValidatingWebhook.
It's not that easy for ValidatingWebhook itself to tell whether a name is auto-generated by MutatingWebhook, like it may require another hash process.
Also if we provide a general response from ValidatingWebhook, users might feel confused if they find their rule names are unique.
I'm glad to hear other better solutions!

Reply:
I don't feel strongly about this. I was advocating for simpler logic in the mutating webhook, but I hear your point and I am fine with the current solution.

Reply:
although i feel strongly about this :D